### ISSUE TITLE
Sketcher: Merging vertices by distance

### ISSUE THREAD

**Issue by kaiwas** [FreeCAD/FreeCAD#16254]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

At the moment, there is no way to mass merge close vertices in the sketch editing mode. We are forced to manually select pairs using single clicks.

But if there was a tool that would allow you to merge the selection using a frame or all vertices at once within a given radius, it would be convenient.

![Merge by distance](https://github.com/user-attachments/assets/331bb452-bdcd-4949-b0e4-d61807a0441c)

An example of work from blender. This often helps out.

https://github.com/user-attachments/assets/e08573c4-34ea-4c6b-a015-1a26f0b083f3

### Full version info

```shell
OS: Manjaro Linux (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38622 (Git) AppImage
Build type: Release
Branch: main
Hash: ed97f3c4591303ee3b54906768f8f4eba5211265
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: Russian/Russia (ru_RU)
Installed mods: 
  * OpenTheme 2024.9.1
  * PitchedRoof
```

### Subproject(s) affected?

Sketcher

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by FEA-eng** [comment]
The [ValidateSketch](https://wiki.freecad.org/Sketcher_ValidateSketch) tool allows you to find and automatically fix (what you are requesting here) missing coincidences within a specified distance tolerance. But having it available while editing a sketch would indeed be nice.

**Comment by kaiwas** [comment]
> The [ValidateSketch](https://wiki.freecad.org/Sketcher_ValidateSketch) tool allows you to find and automatically fix (what you are requesting here) missing coincidences within a specified distance tolerance. But having it available while editing a sketch would indeed be nice.

Yes, I know about this possibility. Thanks. But it would be interesting to move this function to another level.
Also, you can add "only selected" to this.

**Comment by JustTryingToGetSomeWorkDone** [comment]
Yes please on this feature. I just wasted 20 minutes using validate sketch and trying to fix vertices that seemed to have coincidence but somehow still managed to have open wires. I just had to keep deleting the same line segments and reconnecting in different ways, then go back and try validate sketch (always with the coincidence icon popping up over the vertex while clicking) until it somehow finally worked. 

The highlight troublesome vertices function in validate sketch kept finding the same two vertices, but even when deleting length constraints and moving wires, they all moved together. I zoomed as far as it would allow and they still looked like only one vertex, so, without something to automatically repair this, (yes I clicked on the find/fix buttons in validate sketch, and yes I clicked on all of them) time will forever be wasted with this problem. 

Moving validate sketch options into the sketch window would certainly be more convenient than having to close the sketch, try the pad, pocket,  or whatever in part design, then go back to sketcher and open validate sketch before finally going back in and editing the sketch (and then doing it all over again and again). Too many bench changes and clicks.

Maybe the Find missing coincidences function has a bug? Maybe there's a bug in the closed wire code? That was the most frustration I've had in FreeCAD since before v1 was released. 

V1 is still so good. I love FreeCAD, don't get me wrong, but something still needs to be done about manifold issues.

Also great job kaiwas on those modified screenshots showing what you're suggesting. That's very nice effort put into a feature request.

**Comment by MikePryadko** [comment]
@JustTryingToGetSomeWorkDone Hi! Can you reproduce problem you describe? I think example file can really helps to developers.

**Comment by kaiwas** [comment]
@JustTryingToGetSomeWorkDone  look this https://github.com/FreeCAD/FreeCAD/issues/24335

**Comment by JustTryingToGetSomeWorkDone** [comment]
I'll see if I can find the file and get the problem to reappear, but I can't make any promises. I don't know why it happened the first time.

**Comment by Detective-Tapp** [comment]
@MikePryadko Hello! I've come to this thread because I was searching if Freecad had a tool similar too Blenders merge vertices.
I am still learning and fumbling about, transitioning from Fusion360 to not be dependend on Windows, and in my current project I've gotten stuck on an almost infinite amount of open wires. It started off as two, but after deleting duplicate sketch lines (I don't even know how, probably the cutting tool) and then re-constraining the vertices the number has only increased.

Anyways, I could share that file if there is interest and I get shown something that says how.

**Comment by JustTryingToGetSomeWorkDone** [comment]
> [@MikePryadko](https://github.com/MikePryadko) Hello! I've come to this thread because I was searching if Freecad had a tool similar too Blenders merge vertices. I am still learning and fumbling about, transitioning from Fusion360 to not be dependend on Windows, and in my current project I've gotten stuck on an almost infinite amount of open wires. It started off as two, but after deleting duplicate sketch lines (I don't even know how, probably the cutting tool) and then re-constraining the vertices the number has only increased.
> 
> Anyways, I could share that file if there is interest and I get shown something that says how.

Just open a file manager window to your FreeCAD project file, right click on the save file and compress/zip it,  and when you reply in this issue thread,  just drag the zip file onto the text input field where you type your comments. If you look Just below the text input field on the left side, you'll see a paper clip and it says, "Paste, drop, or click to add files."  You could click it and browse to the file location from there if you find that easier.

**Comment by kaiwas** [comment]
> It started off as two, but after deleting duplicate sketch lines (I don't even know how, probably the cutting tool) and then re-constraining the vertices the number has only increased.

You can partially solve your problem as shown in the video.
Check the sketch in object mode.
But! This doesn't always work; you have to exit edit mode and there's no visual control.

https://github.com/user-attachments/assets/54432d5a-3352-4a81-81ea-178213fd4799

**Comment by Detective-Tapp** [comment]
Hey! I guess apologies are in order, cause with a fresh mind I fixed it lickety split. 
There were only two remaining duplicate lines. Which were found and fixed quickly by searching the sketch for black vertices.

You'll have to take my word for it, but the whole top left part of my sketch, 

<img width="696" height="722" alt="Image" src="https://github.com/user-attachments/assets/d88b6ff5-6333-4845-959c-5a977e8b996c" />


had duplicate geometry for every line (even though most vertices showed up as red).  I assume cuase the red vertices take priority over black ones when they are occupying the same space?
This meant that the only way for me to find the problem geometry was to box select and look in the tree what was being highlighted.
And if you wanna know what my sketching process was, I first made that hook shape, after which I offset a couple of sections of it because it wouldn't go in one. This gave me multiple slots? which I then snipped up with the "Trim Edge" tool removing all excess lines.

It is very sad that I don't have a save at the beginning of my troubles, I tried the FCbak, but that one is at 3 duplicate lines.
Also, trying to go further back in the project made me realise that the ctrl+Z history does not transfer on sessions.
Any plans to add something like that?

Anyways, Thanks for the responses. 
I do still hope that the merge vertices tool gets some thought,
even if my situation that led to my need of it was probably user error.

[Tape dispenser.zip](https://github.com/user-attachments/files/27163230/Tape.dispenser.zip)
[Tape dispenser bak.zip](https://github.com/user-attachments/files/27163882/Tape.dispenser.bak.zip)
